ocs_ci/utility/assisted_installer.py

```diff
@@ -0,0 +1,517 @@
+        """
+        Initialize required variables
+
+        Args:
+            environment (str): Environment defined in configuration (default: production)
+
+        """
+        # TODO: check required configuration
+        self.env = environment
+
+        self._token = None
+        self._token_exp = 0
+
+        # Load configuration
+        # SSO URL for generating API token
+        self.sso_url = config.AUTH["assisted_installer"][self.env]["sso_url"]
+        # Offline token used for generating API token
+        self.offline_token = config.AUTH["assisted_installer"][self.env][
+            "offline_token"
+        ]
+        # Assisted Installer Console API url
+        self.api_host_url = config.AUTH["assisted_installer"][self.env]["api_url"]
+
+    @property
+    def token(self):
+        """
+        Property for obtaining API token (based on OFFLINE_TOKEN)
+        """
+        if time.time() > (self._token_exp - 60):
+            logger.debug("Refreshing API token")
+            full_sso_url = urljoin(
+                self.sso_url,
+                "/auth/realms/redhat-external/protocol/openid-connect/token",
+            )
+            headers = {
+                "Accept": "application/json",
+                "Content-Type": "application/x-www-form-urlencoded",
+            }
+            data = {
+                "grant_type": "refresh_token",
+                "client_id": "cloud-services",
+                "refresh_token": self.offline_token,
+            }
+            now = time.time()
+            resp = requests.post(full_sso_url, data=data, headers=headers)
+            resp_json = resp.json()
+            self._token = resp_json["access_token"]
+            self._token_exp = now + resp_json["expires_in"]
+            logger.debug(
+                f"API token refreshed (expires in {self._token_exp - time.time(): .0f} seconds)"
+            )
+        else:
+            logger.debug(
+                f"API token is valid (expires in {self._token_exp - time.time(): .0f} seconds)"
+            )
+
+        return self._token
+
+    @property
+    def api_base_url(self):
+        """
+        Prepare and return base URL for API
+        """
+        return urljoin(self.api_host_url, f"api/{self.api_section}/{self.api_version}/")
+
+    @property
+    def headers(self):
+        """
+        prepare headers for API requests
+        """
+        headers = {
+            "Authorization": f"Bearer {self.token}",
+            "Content-Type": "application/json",
+        }
+        return headers
+
+    def get_api_endpoint_url(self, endpoint=""):
+        """
+        prepare and return full api endpoint URL
+
+        Args:
+            endpoint (str): the final part of the endpoint URL
+
+        Returns:
+            str: full API endpoint URL
+        """
+        return urljoin(self.api_base_url, endpoint)
+
+    def get_request(self, endpoint, params=None, json=True, ignore_failure=False):
```

Comment:
I would suggest to add a retry decorator to all "_request" functions because of the bad connections and bad responses

Reply:
Good point. I didn't faced any issue with accessing the API yet though, so not sure, if I should add the _retry_ right now or keep it without that, so we will know about potential issues and we could debug it if it is just network issue where the _retry_ makes sense or if it is another problem which we should fix differently. WDYT?